### PR TITLE
New features for pathway identification and charge barrier analysis

### PR DESCRIPTION
**Pull request by jmmshn**
# New features for pathway identification and charge barrier analysis
- Added method for performing Dijkstra-like pathway analysis on periodic graph.
- Added the functionality to find a sequence of hops that take you from site-i to periodic image of site-i thus and identify intercalating pathways
- Added functionality to assign a cost to individual migration events either using the integrated charge density along a tube or along s series of image sites determined by the PathFinder tool from pymatgen.

### CODE REVIEW

**Comment by jmmshn** [comment]
Hi @shyuep, apologies for how this large this PR is.  Much of this we kept in a private repo over last year for internal projects. Will it be possible to give me write access?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+repos:
```
Comment (shyuep) [inline .pre-commit-config.yaml:1]: I am unclear what this file is for or why it needs to be committed?

File: pymatgen_diffusion/neb/full_path_mapper.py [hunk 2]
```diff
@@ -62,16 +82,19 @@ class FullPathMapper(MSONable):
...
-                 vac_mode=False):
+    def __init__(
+        self,
+        structure,
+        migrating_specie,
+        max_path_length=10,
+        symprec=0.1,
+        vac_mode=False,
+        name: str = None,
+    ):
```
Comment (shyuep) [inline pymatgen_diffusion/neb/full_path_mapper.py:97]: Some functional annotations would be useful.

File: pymatgen_diffusion/neb/full_path_mapper.py [hunk 3]
```diff
@@ -129,61 +161,233 @@ def _get_pos_and_migration_path(self, u, v, w):
...
+
+        # for u, v, k, d in self.s_graph.graph.edges(data=True, keys=True):
+        for u in self.s_graph.graph.nodes():
+            # for each hop that leave the UC cut that hop and get the path to other cell
+            # if d["to_jimage"] == (0, 0, 0):
+            #     continue
+            # if d["cost"] > max_val:
+            #     continue
+            # Create a copy of the graph that is only composed of internal hops
+            GG = deepcopy(self.s_graph.graph)
```
Comment (shyuep) [inline pymatgen_diffusion/neb/full_path_mapper.py:282]: This does not seem to follow python style guides.

File: setup.cfg [hunk 4]
```diff
@@ -1,6 +1,12 @@
 [pycodestyle]
 count = True
-ignore=E121,E123,E126,E133,E226,E241,E242,E704,W503,W504,W505,E741,W605
+ignore=E121,E123,E126,E133,E226,E241,E242,E704,W503,W504,W505,E741,W605,E203
```
Comment (shyuep) [inline setup.cfg:3]: What is E203?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jmmshn) [reply .pre-commit-config.yaml:1]: This file is to help format the changed files prior to committing
Committing this ensures that the code changes will pass most linting checks after a push.
We use this for pymatgen:
https://github.com/materialsproject/pymatgen/blob/master/.pre-commit-config.yaml

If you do not install the hook it does nothing but if it's installed it's just a nice little creature comfort thing for developers.

File: setup.cfg (hunk 4 above)
Comment (jmmshn) [reply setup.cfg:3]: black (which is called by the pre-commit stack defined in the yaml file above) reformat the white space around ":" correctly but pycodetyle returns a false positive.

https://github.com/psf/black/issues/279
https://github.com/PyCQA/pycodestyle/issues/373

File: pymatgen_diffusion/neb/full_path_mapper.py (hunk 3 above)
Comment (jmmshn) [reply pymatgen_diffusion/neb/full_path_mapper.py:282]: sorry I clean up the comments around here and gave it a better name too

File: pymatgen_diffusion/neb/full_path_mapper.py (hunk 2 above)
Comment (jmmshn) [reply pymatgen_diffusion/neb/full_path_mapper.py:97]: added

**Comment by jmmshn** [comment]
@shyuep  I think this is ready for merge now.

**Comment by shyuep** [comment]
Thanks.

**Pull request merged**